lib/Heartbeat.js

```diff
@@ -0,0 +1,248 @@
+    frag.appendChild(rightSpacer);
+
+    // collapse the space before the stars
+    this.messageText.flex = 0;
+    let leftSpacer = this.messageText.nextSibling;
+    leftSpacer.flex = 0;
+
+    // Append the fragment and apply the styling
+    this.notice.appendChild(frag);
+    this.notice.classList.add('heartbeat');
+
+    // Let the consumer know the notification was shown.
+    this.maybeNotifyHeartbeat('NotificationOffered');
+
+    let handleWindowClosed = () => {
+      this.maybeNotifyHeartbeat('WindowClosed');
+      this.chromeWindow.removeEventListener('SSWindowClosing', handleWindowClosed);
+    };
+    this.chromeWindow.addEventListener('SSWindowClosing', handleWindowClosed);
+
+    let surveyDuration = Services.prefs.getIntPref(PREF_SURVEY_DURATION) * 1000;
+    this.surveyEndTimer = setTimeout(() => {
+      this.maybeNotifyHeartbeat('SurveyExpired');
+      this.notificationBox.removeNotification(this.notice);
+    }, surveyDuration);
+  }
+
+  maybeNotifyHeartbeat(name, data={}) {
+    Log.log('heartbeat event', name, data);
+
+    if (this.pingSent) {
+      Log.warn('Heartbeat event recieved after Telemetry ping sent. name:', name, 'data:', data);
+    }
+
+    let timestamp = Date.now();
+    let sendPing = false;
+
+    switch (name) {
+      case 'NotificationOffered':
+        this.surveyResults.flowId = this.options.flowId;
+        this.surveyResults.offeredTS = timestamp;
+        break;
+      case 'LearnMore':
+        if (!this.surveyResults.learnMoreTS) {
+          this.surveyResults.learnMoreTS = timestamp;
+        }
+        break;
+      case 'Engaged':
+        this.surveyResults.engagedTS = timestamp;
+        break;
+      case 'Voted':
+        this.surveyResults.votedTS = timestamp;
+        this.surveyResults.score = data.score;
+        break;
+      case 'SurveyExpired':
+        this.surveyResults.expiredTS = timestamp;
+        break;
+      case 'NotificationClosed':
+        this.surveyResults.closedTS = timestamp;
+        sendPing = true;
+        break;
+      case 'WindowClosed':
+        this.surveyResults.windowClosedTS = timestamp;
+        sendPing = true;
+        break;
+      default:
+        Log.error('Unrecognized Heartbeat event:', name);
+        break;
+    }
+
+    data.timestamp = timestamp;
+    data.flowId = this.options.flowId;
+    this.events.emit(name, data);
+
+    if (!sendPing) {
+      return;
+    }
+
+    // Send the ping to Telemetry
+    let payload = Object.assign({version: 1}, this.surveyResults);
+    for (let meta of ['surveyId', 'surveyVersion', 'testing']) {
+      if (this.options.hasOwnProperty(meta)) {
+        payload[meta] = this.options[meta];
+      }
+    }
+
+    Log.debug('Sending payload to Telemetry: name:', name, 'payload:', payload);
+
+    TelemetryController.submitExternalPing('heartbeat', payload, {
+      addClientId: true,
+      addEnvironment: true,
+    });
+
+    // only for testing
+    this.events.emit('TelemetrySent', payload);
```

Comment:
Should we remove this for production? I'm assuming this is just for logging purposes?